libs/serial/serial.ts

```diff
@@ -29,28 +39,50 @@ namespace serial {
     //% weight=20 blockGap=8
     //% group="Read"
     export function readLine(): string {
-        return serial.readUntil("\n");
+        return serial.readUntil(Delimiters.NewLine);
     }
 
     /**
-     * Return the corresponding delimiter string
+     * Read a line of text from the serial port and return the buffer when the delimiter is met.
+     * @param delimiter text delimiter that separates each text chunk
      */
-    //% blockId="serial_delimiter_conv" block="%del"
-    //% weight=1 blockHidden=true
-    //% group="Read"
-    export function delimiters(del: Delimiters): string {
-        // even though it might not look like, this is more
-        // (memory) efficient than the C++ implementation, because the
-        // strings are statically allocated and take no RAM
-        switch (del) {
-            case Delimiters.NewLine: return "\n"
-            case Delimiters.Comma: return ","
-            case Delimiters.Dollar: return "$"
-            case Delimiters.Colon: return ":"
-            case Delimiters.Fullstop: return "."
-            case Delimiters.Hash: return "#"
-            default: return "\n"
-        }
+    //% help=serial/read-until
+    //% blockId=serial_read_until block="serial|read until %delimiter=serial_delimiter_conv"
+    //% weight=19
+    //% group="Read"    
+    export function readUntil(delimiter: Delimiters, timeOut?: number): string {
+        const start = control.millis();
+        let r = "";
+        let buf = control.createBuffer(3);
+        let bufi = 0;
+        do {
+            const c = serial.read();
+            if (c < 0) // error -- return what we have so far
+                break;
+            else if (bufi == 0 && c == delimiter) // found it!                
+                break;
```

Comment:
Skip bufi==0 - if it's not the case you had a malformed char, which should be OK to drop - delimiters have the high bit clear whereas multi byte chars have it set

Reply:
move the check to ascii case